packages/flutter/lib/src/widgets/localizations.dart

```diff
@@ -0,0 +1,465 @@
+  ///
+  /// Typically defined with [DefaultLocalizationsDelegate].
+  final LocalizationsDelegate delegate;
+
+  /// The widget below this widget in the tree.
+  final Widget child;
+
+  /// The locale of the Localizations widget for the widget tree that
+  /// corresponds to [BuildContext] `context`.
+  static Locale localeOf(BuildContext context) {
+    assert(context != null);
+    final _LocalizationsScope scope = context.inheritFromWidgetOfExactType(_LocalizationsScope);
+    return scope.localizationsState.locale;
+  }
+
+  /// Returns the 'type' localized resources for the widget tree that
+  /// corresponds to [BuildContext] `context`.
+  ///
+  /// This method is typically used by a static factory method on the 'type'
+  /// class. For example Flutter's MaterialLocalizations class looks up Material
+  /// resources with a method defined like this:
+  ///
+  /// ```dart
+  /// static MaterialLocalizations of(BuildContext context) {
+  ///    return Localizations.of<MaterialLocalizations>(context, MaterialLocalizations);
+  /// }
+  /// ```
+  static T of<T>(BuildContext context, Type type) {
+    assert(context != null);
+    assert(type != null);
+    final _LocalizationsScope scope = context.inheritFromWidgetOfExactType(_LocalizationsScope);
+    return scope.localizationsState.resourcesFor<T>(type);
+  }
+
+  @override
+  _LocalizationsState createState() => new _LocalizationsState();
+}
+
+class _LocalizationsState extends State<Localizations> {
+  final GlobalKey _localizedResourcesScopeKey = new GlobalKey();
+
+  Locale get locale => _locale;
+  Locale _locale;
+
+  @override
+  void initState() {
+    super.initState();
+    if (widget.delegate == null)
+      _locale = widget.locale;
+    else
+      load(widget.locale);
+  }
+
+  @override
+  void didUpdateWidget(Localizations old) {
+    super.didUpdateWidget(old);
+    if (widget.delegate != old.delegate && widget.delegate != null)
+      load(widget.locale);
+  }
+
+  void load(Locale locale) {
+    final LocalizationsDelegate delegate = widget.delegate;
+    if (delegate == null) {
+      _locale = locale;
+      return;
+    }
+
+    Iterable<Type> types;
+    final Future<Iterable<Type>> futureTypes = delegate.load(locale)
+      .then<Iterable<Type>>((Iterable<Type> value) {
+        return types = value;
+      });
+
+    if (types != null) {
+      // The delegate loaded its resources synchronously, do not rebuild
+      _locale = locale;
+    } else {
+      // - Don't rebuild the dependent widgets until the resources for the new locale
+      // have finished loading. Until then the old locale will continue to be used.
+      // - If we're running at app startup time then defer reporting the first
+      // "useful" frame until after the async load has completed.
+      WidgetsBinding.instance.deferFirstFrameReport();
+      futureTypes.then((Iterable<Type> value) {
+        WidgetsBinding.instance.allowFirstFrameReport();
```

Comment:
It seems scary that even if the Localizations widget is removed from the tree, the binding won't report the first frame until this future completes (which might never happen).